common/models/transducer/transducer_training_pipeline/fixed_path_training.py

```diff
@@ -0,0 +1,138 @@
+from __future__ import annotations
+from typing import Dict, Any
+from returnn.tf.util.data import DimensionTag
+from returnn.config import get_global_config
+from ....training.switchout import switchout_target
+from .pipeline import Context
+
+
+def update_net_for_fixed_path_training(net: Dict[str, Any],
+                                       ctx: Context,
+                                       stage_num_align: int,
+                                       align_dir: str,
+                                       switchout: bool = True,
+                                       decoder: str = "output",
+                                       inited_switchout_output: str = "output_",
+                                       output_log_prob_wb: str = "output_log_prob_wb"
+                                       ) -> Dict[str, Any]:
+  """
+  Args:
+      net (Dict[str, Any]): [description]
+      ctx (Context): context providing us with extra info.
+      stage_num_align (int): Stage number from which we take the dumped alignments.
+      align_dir (str): Path to the folder with the .hdf files.
+      switchout (bool): Whether to do switchout on the predicted labels.
+      decoder (str): Whether to do switchout on the predicted label.
+      inited_switchout_output (str): from this output_is_not_blank is calculated.
+
+  Returns:
+      Dict[str, Any]: [updated network dictionary]
+  """
+  new_target_name = "targetb"
+  target = ctx.target
+  blank_idx = ctx.blank_idx
+  train = ctx.train
+
+  subnet = net[decoder]["unit"]
+  subnet["target"] = new_target_name
+
+  if train:
+    subnet["size_target"] = new_target_name
+    del subnet["lm_input"]
+    del subnet["full_sum_loss"]
+    if switchout:  # Framewise training allows switchout
+      subnet[inited_switchout_output] = {  # SwitchOut in training
+        "class": "eval", "from": "output", "eval": switchout_target,
+        "eval_local": {"targetb_blank_idx": blank_idx, "target_num_labels": target.get_num_classes()},
+        "initial_output": 0
+      }
+
+  # The layer name must be smaller than "t_target" such that this is created first.
+  net["existing_alignment"] = {"class": "reinterpret_data",
+                               "from": "data:alignment",
+                               "set_sparse_dim": target.get_num_classes(),
+                               "size_base": "encoder",  # TODO: for RNA only...
```

Comment:
any idea how it can be set for RNNT? 
I.e. how to give there the sum of the sizes of encoder and decoder
@albertz

Reply:
I thought we have an example for this somewhere?

Reply:
For the RNNT, Andre has it out-commented. He skips the `reinterpret_data` layer and just continues with:
``` python
    "1_targetb_base": {
        "class": "copy",
        # "from": "existing_alignment",
        "from": "data:alignment",
        "register_as_extern_data": "targetb" if task == "train" else None},
```
Otherwise I couldn't find an other example.

So the question is if we really need the reinterpret layer?

Reply:
That's not what I meant. I meant for the `size_base`. But maybe that's not needed (you need it when you want to tell RETURNN that it is the same to some other dim tag).

I also see that it sets the sparse dim. Although that looks a bit incorrect anyway. It would include Blank Labels at this point, and I think `target` here is without Blank. Only once you remove the Blank Frames/Labels, this makes sense. But maybe this is also not relevant (depending on how it is used).

Note that RNNT Training with fixed alignment is anyway not fully supported yet, because chunking doesn't fully work. (See [here](https://github.com/rwth-i6/returnn/pull/375).)

Reply:
So the question is, do we need to tell RETURNN about the shape of the targets if we are training with framewise CE? Or when should RETURNN know about the shape of the targets?

>Although that looks a bit incorrect anyway. It would include Blank Labels at this point, and I think target here is without Blank.

Nice that you caught that one. It should have had the dim inclusive the blank.

 I have looked into the issue with chunking. Will first commit it like this and integrate it later. The possibilities are, it either can be solved from [#376](https://github.com/rwth-i6/returnn/issues/376) or through the workaround of Andre.  
I see that he used it in some of his configs but am not sure how he worked the following problem out(or didn't at all):

> But even that is hacky and ugly, and will break in some cases, e.g. when you define custom_iterate_seqs in some epochs, and later not anymore. Then it would not correctly reset this.